sysl2/grammarParser.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+    sysl "anz-bank/sysl/sysl2/proto"
+    "sort"
+)
+
+type parser struct {
+    g         *sysl.Grammar
+    l         *lexer
+    terminals []string
+}
+
+type builder struct {
+    arr    []string
+    tokens map[string]int32
+    index  int32
+}
+
+const FIRST = 0
+const EPSILON = -2
+const EOF = -1
+
+func (b *builder) handleChoice(choice *sysl.Choice) {
+    for _, s := range choice.Sequence {
+        for _, t := range s.Term {
+            if t == nil {
+                continue
+            }
+            var str string
+            switch t.Atom.Union.(type) {
+            case *sysl.Atom_String_:
+                str = t.GetAtom().GetString_()
+            case *sysl.Atom_Regexp:
+                str = t.GetAtom().GetRegexp()
+                str = `\A(?:` + str + `)`
+            case *sysl.Atom_Choices:
+                b.handleChoice(t.GetAtom().GetChoices())
+            }
+            if str != "" {
+                _, has := b.tokens[str]
+                if !has {
+                    b.tokens[str] = b.index
+                    b.arr = append(b.arr, str)
+                    t.Atom.Id = b.index
```

Comment:
No need for `index`. Use `len(b.arr)` instead.

Reply:
done.